In [1]:
#Import Libraries
import numpy as np
import pandas as pd
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LogisticRegression
from sklearn.neighbors import KNeighborsClassifier
from sklearn.tree import DecisionTreeClassifier
from sklearn.metrics import accuracy_score, confusion_matrix

In [3]:
df = pd.read_csv("Iris.csv")
df.head()

Id  SepalLengthCm  SepalWidthCm  PetalLengthCm  PetalWidthCm      Species
0   1            5.1           3.5            1.4           0.2  Iris-setosa
1   2            4.9           3.0            1.4           0.2  Iris-setosa
2   3            4.7           3.2            1.3           0.2  Iris-setosa
3   4            4.6           3.1            1.5           0.2  Iris-setosa
4   5            5.0           3.6            1.4           0.2  Iris-setosa

In [4]:
# First Row is just and ID which is not considered as a feature here
Input = df.iloc[:,1:4].values
Output = df.iloc[:,5].values

print(Input.shape, "\n", Output.shape)

(150, 3) 
 (150,)


In [5]:
X_train, X_test, Y_train,Y_test = train_test_split(Input, Output, test_size=0.15, random_state=0)
print("\n", X_train.shape, "\n", Y_train.shape)
print("\n", X_test.shape, "\n", Y_test.shape)


 (127, 3) 
 (127,)

 (23, 3) 
 (23,)


In [5]:
#Note: here out data set value ranges between different features are within comparable scale (cm)
#So we wont be needing the Standard scaler for here
#Here we have more than two class, generaly logistic regression is binary classification
#Thus the Function does one to many regression,i.e, if there is N classes, 
#there would be N binary logistic regresion to check whether the sample is belonging to that class or not

log = LogisticRegression()
log.fit(X_train, Y_train)
Y_pred = log.predict(X_test)

print("\nAccuracy:", accuracy_score(Y_test, Y_pred)*100, "%")
print("\nPrediction / Actual results\n")
print(np.concatenate((Y_pred.reshape(-1,1), Y_test.reshape(-1,1)), axis =1))
print("\n Confusion Matrix: ")
print(confusion_matrix(Y_test, Y_pred))


Accuracy: 100.0 %

Prediction / Actual results

[['Iris-virginica' 'Iris-virginica']
 ['Iris-versicolor' 'Iris-versicolor']
 ['Iris-setosa' 'Iris-setosa']
 ['Iris-virginica' 'Iris-virginica']
 ['Iris-setosa' 'Iris-setosa']
 ['Iris-virginica' 'Iris-virginica']
 ['Iris-setosa' 'Iris-setosa']
 ['Iris-versicolor' 'Iris-versicolor']
 ['Iris-versicolor' 'Iris-versicolor']
 ['Iris-versicolor' 'Iris-versicolor']
 ['Iris-virginica' 'Iris-virginica']
 ['Iris-versicolor' 'Iris-versicolor']
 ['Iris-versicolor' 'Iris-versicolor']
 ['Iris-versicolor' 'Iris-versicolor']
 ['Iris-versicolor' 'Iris-versicolor']
 ['Iris-setosa' 'Iris-setosa']
 ['Iris-versicolor' 'Iris-versicolor']
 ['Iris-versicolor' 'Iris-versicolor']
 ['Iris-setosa' 'Iris-setosa']
 ['Iris-setosa' 'Iris-setosa']
 ['Iris-virginica' 'Iris-virginica']
 ['Iris-versicolor' 'Iris-versicolor']
 ['Iris-setosa' 'Iris-setosa']]

 Confusion Matrix: 
[[ 7  0  0]
 [ 0 11  0]
 [ 0  0  5]]


In [6]:
#Note: here the algorithim's accuracy is pure based on the K-Neighbor value
#The best accuracy is usually in the middle not too high or not too low, so it needs to be found out by hit and trial
#Mostly the K value cant be greater than no.of classes (here 3)
#hence here i take a loop from 1 to 5, finding which has the highest accuracy
acc = 0
K = 0
for i in range(1,6):    
    knn = KNeighborsClassifier(n_neighbors=i)
    knn.fit(X_train, Y_train)
    Y_pred = knn.predict(X_test)
    
    print("K-Neighbhor:", i, "\t\tAccuracy:", accuracy_score(Y_test, Y_pred)*100, "%")
    
    if(accuracy_score(Y_test, Y_pred)*100 > acc):
        acc = accuracy_score(Y_test, Y_pred)*100
        Y_pred_final = Y_pred
        K = i

print("\nK-Neighbhor:", K, "\t\tAccuracy:", accuracy_score(Y_test, Y_pred_final)*100, "%")
print("\n Prediction / Actual result\n")
print(np.concatenate((Y_pred_final.reshape(-1,1), Y_test.reshape(-1,1)), axis =1))
print("\n Confusion Matrix: ")
print(confusion_matrix(Y_test, Y_pred))

K-Neighbhor: 1 		Accuracy: 95.65217391304348 %
K-Neighbhor: 2 		Accuracy: 100.0 %
K-Neighbhor: 3 		Accuracy: 91.30434782608695 %
K-Neighbhor: 4 		Accuracy: 91.30434782608695 %
K-Neighbhor: 5 		Accuracy: 91.30434782608695 %

K-Neighbhor: 2 		Accuracy: 100.0 %

 Prediction / Actual result

[['Iris-virginica' 'Iris-virginica']
 ['Iris-versicolor' 'Iris-versicolor']
 ['Iris-setosa' 'Iris-setosa']
 ['Iris-virginica' 'Iris-virginica']
 ['Iris-setosa' 'Iris-setosa']
 ['Iris-virginica' 'Iris-virginica']
 ['Iris-setosa' 'Iris-setosa']
 ['Iris-versicolor' 'Iris-versicolor']
 ['Iris-versicolor' 'Iris-versicolor']
 ['Iris-versicolor' 'Iris-versicolor']
 ['Iris-virginica' 'Iris-virginica']
 ['Iris-versicolor' 'Iris-versicolor']
 ['Iris-versicolor' 'Iris-versicolor']
 ['Iris-versicolor' 'Iris-versicolor']
 ['Iris-versicolor' 'Iris-versicolor']
 ['Iris-setosa' 'Iris-setosa']
 ['Iris-versicolor' 'Iris-versicolor']
 ['Iris-versicolor' 'Iris-versicolor']
 ['Iris-setosa' 'Iris-setosa']
 ['Iris-setosa' 'I

In [6]:
#Decision Tree

dt = DecisionTreeClassifier()
dt.fit(X_train, Y_train)
Y_pred = dt.predict(X_test)

print("\nAccuracy:", accuracy_score(Y_test, Y_pred)*100, "%")
print("\nPrediction / Actual results\n")
print(np.concatenate((Y_pred.reshape(-1,1), Y_test.reshape(-1,1)), axis =1))
print("\n Confusion Matrix: ")
print(confusion_matrix(Y_test, Y_pred))


Accuracy: 95.65217391304348 %

Prediction / Actual results

[['Iris-virginica' 'Iris-virginica']
 ['Iris-versicolor' 'Iris-versicolor']
 ['Iris-setosa' 'Iris-setosa']
 ['Iris-virginica' 'Iris-virginica']
 ['Iris-setosa' 'Iris-setosa']
 ['Iris-virginica' 'Iris-virginica']
 ['Iris-setosa' 'Iris-setosa']
 ['Iris-versicolor' 'Iris-versicolor']
 ['Iris-versicolor' 'Iris-versicolor']
 ['Iris-versicolor' 'Iris-versicolor']
 ['Iris-versicolor' 'Iris-virginica']
 ['Iris-versicolor' 'Iris-versicolor']
 ['Iris-versicolor' 'Iris-versicolor']
 ['Iris-versicolor' 'Iris-versicolor']
 ['Iris-versicolor' 'Iris-versicolor']
 ['Iris-setosa' 'Iris-setosa']
 ['Iris-versicolor' 'Iris-versicolor']
 ['Iris-versicolor' 'Iris-versicolor']
 ['Iris-setosa' 'Iris-setosa']
 ['Iris-setosa' 'Iris-setosa']
 ['Iris-virginica' 'Iris-virginica']
 ['Iris-versicolor' 'Iris-versicolor']
 ['Iris-setosa' 'Iris-setosa']]

 Confusion Matrix: 
[[ 7  0  0]
 [ 0 11  0]
 [ 0  1  4]]


-------------------------------------------------------------
_**Finally we can see the ACCURACY of each Algorithim here:**_

--------------------------------------------------------------
Logistic Regression: 100%
<br>KNN: 100%
<br>Decision Tree: 95.65%